From 02b622089bf871c03036c11be9c8fed740b1ad1a Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:05 -0400
Subject: [ethernet] i40e: workaround NVM GLQF_HKEY

Message-id: <1409658042-22689-185-git-send-email-sassmann@redhat.com>
Patchwork-id: 90381
O-Subject: [RHEL7.1 PATCH 184/281] i40e: workaround NVM GLQF_HKEY
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jesse Brandeburg <jesse.brandeburg@intel.com>

The NVM wasn't filling in the GLQF_HKEY register on
some old NVM versions.  If this is the case, fill
in some values so receive with flow rules works right.

Change-ID: Ic737888ee68f96efb4cf8a1a49d2301615e09ed2
Signed-off-by: Jesse Brandeburg <jesse.brandeburg@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 407e063c92c51420c3e919203de47e144dec6934)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index d2818ab..3187444 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -5463,6 +5463,20 @@ static void i40e_fdir_sb_setup(struct i40e_pf *pf)
  struct i40e_vsi *vsi;
  int i;
 
+ /* quick workaround for an NVM issue that leaves a critical register
+  * uninitialized
+  */
+ if (!rd32(&pf->hw, I40E_GLQF_HKEY(0))) {
+  static const u32 hkey[] = {
+   0xe640d33f, 0xcdfe98ab, 0x73fa7161, 0x0d7a7d36,
+   0xeacb7d61, 0xaa4f05b6, 0x9c5c89ed, 0xfc425ddb,
+   0xa4654832, 0xfc7461d4, 0x8f827619, 0xf5c63c21,
+   0x95b3a76d};
+
+  for (i = 0; i <= I40E_GLQF_HKEY_MAX_INDEX; i++)
+   wr32(&pf->hw, I40E_GLQF_HKEY(i), hkey[i]);
+ }
+
  if (!(pf->flags & I40E_FLAG_FD_SB_ENABLED))
   return;
 
-- 
1.7.1